src/combinatorics/catalan.lean

```diff
@@ -0,0 +1,158 @@
+/-
+Copyright (c) 2022 Julian Kuelshammer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Julian Kuelshammer
+-/
+import data.nat.choose.central
+import algebra.big_operators.fin
+import tactic.field_simp
+
+/-!
+# Catalan numbers
+
+The Catalan numbers (http://oeis.org/A000108) are probably the most ubiquitous sequence of integers
+in mathematics. They enumerate several important objects like binary trees, Dyck paths, and
+triangulations of convex polygons.
+
+## Main definitions
+
+* `catalan n`: the `n`th Catalan number, defined recursively as
+               `catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)`.
+* `catalan' n`: the `n`th Catalan number, defined explicitly as
+               `catalan' n = 1 / (n + 1) * nat.central_binom n`.
+
+## Main results
+
+* `catalan_defs_agree`: The fact that the two definitions both give the same sequence.
+
+## Implementation details
+
+The proof of the equality of the two definitions follows
+https://math.stackexchange.com/questions/3304415/catalan-numbers-algebraic-proof-of-the-recurrence-relation
+
+## TODO
+
+* Prove that the Catalan numbers enumerate many interesting objects.
+* Generalise to Catalan numbers associated to arbitrary Coxeter groups.
+
+-/
+
+open_locale big_operators
+open finset
+
+/-- The recursive definition of the sequence of Catalan numbers:
+`catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)` -/
+def catalan : ℕ → ℕ
+| 0       := 1
+| (n + 1) := ∑ i : fin n.succ, have _ := i.2, have _ := nat.lt_succ_iff.mpr (n.sub_le i),
+             catalan i * catalan (n - i)
+
+@[simp] lemma catalan_zero : catalan 0 = 1 := by rw catalan
+
+lemma catalan_succ (n : ℕ) : catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i) :=
+by rw catalan
+
+/-- The explicit definition of the sequence of Catalan numbers via the central binomial
+coefficient.-/
+def catalan' (n : ℕ) :  ℚ :=
+1 / (n + 1) * nat.central_binom n
+
+@[simp] lemma catalan'_zero : catalan' 0 = 1 :=
+begin rw catalan', simp, end
+
+/-- A helper sequence that can be used to prove the equality of the recursive and the explicit
+definition using a telescoping sum argument. -/
+private def gosper_catalan (n j : ℕ) : ℚ :=
+nat.central_binom j * nat.central_binom (n - j) * (2 * j - n) / (2 * n * (n + 1))
+
+lemma gosper_trick (n i : ℕ) (h : i ≤ n) :
```

Comment:
```suggestion
lemma gosper_trick {n i : ℕ} (h : i ≤ n) :
```
Also consider giving it a more descriptive name.

Reply:
Also, if `gosper_catalan` is to be private, then this lemma should be private too.